components/net/http_cache.rs

```diff
@@ -0,0 +1,488 @@
+                    return Some(cached_response);
+                }
+            }
+        }
+        None
+    }
+
+    /// https://tools.ietf.org/html/rfc7234#section-4.3.4 Freshening Stored Responses upon Validation.
+    pub fn refresh(&mut self, request: &Request, response: Response, done_chan: &mut DoneChannel) -> Option<Response> {
+        assert!(response.status == Some(StatusCode::NotModified));
+        let entry_key = CacheKey::new(request.clone());
+        if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
+            for cached_resource in cached_resources.iter_mut() {
+                if let Ok(ref mut stored_headers) = cached_resource.metadata.headers.try_lock() {
+                    // Received a response with 304 status code whose url matches a stored resource.
+                    // 1. update the headers of the stored resource.
+                    // 2. return a response, constructed from the updated stored resource.
+                    stored_headers.extend(response.headers.iter());
+                    let mut constructed_response = Response::new(cached_resource.metadata.final_url.clone());
+                    constructed_response.headers = stored_headers.clone();
+                    constructed_response.body = cached_resource.body.clone();
+                    constructed_response.status = cached_resource.status.clone();
+                    constructed_response.raw_status = cached_resource.raw_status.clone();
+                    constructed_response.url_list = cached_resource.url_list.clone();
+                    // done_chan will have been set to Some by http_network_fetch,
+                    // set it back to None since the response returned here replaces the 304 one from the network.
+                    *done_chan = None;
+                    let mut expires = cached_resource.expires.lock().unwrap();
+                    *expires = get_response_expiry(&constructed_response);
+                    return Some(constructed_response);
+                }
+            }
+        }
+        None
+    }
+
+    fn invalidate_for_url(&mut self, url: String) {
+        let entry_key = CacheKey::from_url_string(url);
+        if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
+            for cached_resource in cached_resources.iter_mut() {
+                let mut expires = cached_resource.expires.lock().unwrap();
+                *expires = Duration::seconds(0i64);
+            }
+        }
+    }
+
+    /// https://tools.ietf.org/html/rfc7234#section-4.4 Invalidation.
+    pub fn invalidate(&mut self, request: &Request, response: &Response) {
+        if let Some(status) = response.status {
+            // Ignoring redirects.
+            // Note: without this, the fetch::test_fetch_redirect_updates_method would hang.
+            // An unknown interaction between the cache and redirects caused this problem.
```

Comment:
Ok, I think this was actually a deadlock inside `construct_response`, since `create_cached_response` was locking the headers that were locked just before inside `construct_response`. I'm not sure why this only was apparent inside the revalidation of a re-direct though. 

Fix https://github.com/servo/servo/pull/18676/commits/1cef9cec53934d031b34c378461b25bc5ed79a63

Also removed the check for redirect status inside `invalidate`, unit test passing again without the check. https://github.com/servo/servo/pull/18676/commits/bd10d895dc19052af4940177d34ef89eb09ac655